matter_server/server/device_controller.py

```diff
@@ -0,0 +1,391 @@
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Coroutine,
+    Deque,
+    Final,
+    Generator,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+    cast,
+)
+
+from chip.ChipDeviceCtrl import ChipDeviceController
+from chip.discovery import FilterType as DiscoveryFilterType
+from chip.exceptions import ChipStackError
+
+from matter_server.server.const import SCHEMA_VERSION
+
+from ..common.helpers.api import api_command
+from ..common.helpers.util import dataclass_from_dict, dataclass_to_dict
+from ..common.models.device import CommissionOption
+from ..common.models.error import (
+    NodeCommissionFailed,
+    NodeInterviewFailed,
+    NodeNotExists,
+)
+from ..common.models.events import EventType
+from ..common.models.message import (
+    CommandMessage,
+    ErrorResultMessage,
+    SuccessResultMessage,
+)
+from ..common.models.node import MatterNode
+
+if TYPE_CHECKING:
+    from chip.clusters import (
+        Attribute,
+        Cluster,
+        ClusterAttributeDescriptor,
+        ClusterEvent,
+        ClusterCommand
+    )
+
+    from .server import MatterServer
+    from .stack import MatterStack
+
+
+DATA_KEY_NODES = "nodes"
+DATA_KEY_LAST_NODEID = "last_nodeid"
+
+
+class MatterDeviceController:
+    """Class that manages the Matter devices."""
+
+    def __init__(
+        self,
+        server: MatterServer,
+    ):
+        self.server = server
+        self.logger = server.logger.getChild("device_controller")
+        # Instantiate the underlying ChipDeviceController instance on the Fabric
+        self.chip_controller: ChipDeviceController = (
+            server.stack.fabric_admin.NewController()
+        )
+        # we keep the last events in memory so we can include them in the diagnostics dump
+        self.event_history: Deque[Attribute.EventReadResult] = deque(maxlen=25)
+        self._subscriptions: dict[int, Attribute.SubscriptionTransaction] = {}
+        self._nodes: dict[int, MatterNode] = {}
+        self._wifi_creds_set = False
+        self.logger.debug("CHIP Device Controller Initialized")
+
+    @property
+    def fabric_id(self) -> int:
+        """Return Fabric ID."""
+        return self.chip_controller.fabricId
+
+    @property
+    def compressed_fabric_id(self) -> int:
+        """Return unique identifier for this initialized fabric."""
+        return self.chip_controller.GetCompressedFabricId()
+
+    async def start(self) -> None:
+        """Async initialize of controller."""
+        # load nodes from persistent storage
+        nodes_data = self.server.storage.get(DATA_KEY_NODES, {})
+        for nodeid_str, node_dict in nodes_data.items():
+            nodeid = int(nodeid_str)
+            # TEMP !!!! TODO
+            # node_dict["attributes"] = {}
+            node = dataclass_from_dict(MatterNode, node_dict)
+            self._nodes[nodeid] = node
+            # make sure to start node subscriptions
+            await self._subscribe_node(nodeid)
```

Comment:
```suggestion
        for node_id_str, node_dict in nodes_data.items():
            node_id = int(node_id_str)
            # TEMP !!!! TODO
            # node_dict["attributes"] = {}
            node = dataclass_from_dict(MatterNode, node_dict)
            self._nodes[node_id] = node
            # make sure to start node subscriptions
            await self._subscribe_node(node_id)
```

Reply:
Do we really want to change it to "node_id" if the SDK calls it "nodeid" ? I see this biting us some time ;-) 
From the other hand we should create our own api we feel comfortable with and not mimic another...

Just as with the camel_case vs snake_case we discussed before, I'm in favor of changing it to our own liking

Reply:
I'd prefer always calling it `node_id`. I don't think we'll get confused about what it is.